src/WindowsUtils/Debugger.cpp

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2022 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "WindowsUtils/Debugger.h"
+
+#include "OrbitBase/GetLastError.h"
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/StringConversion.h"
+#include "OrbitBase/ThreadUtils.h"
+#include "WindowsUtils/CreateProcess.h"
+#include "WindowsUtils/SafeHandle.h"
+
+namespace orbit_windows_utils {
+
+Debugger::Debugger() : process_info_or_error_(ErrorMessage()) {}
+Debugger::~Debugger() { Wait(); }
+
+ErrorMessageOr<ProcessInfo> Debugger::Start(const std::filesystem::path& executable,
+                                            const std::filesystem::path& working_directory,
+                                            std::string_view arguments) {
+  // Launch process and start debugging loop on the same separate thread.
+  thread_ = std::thread(&Debugger::DebuggerThread, this, executable, working_directory,
+                        std::string(arguments));
+
+  // We use "create_process_promise_" to wait until the result of "CreateProcessToDebug", which is
+  // stored in "process_info_or_error_", is available from the  debugger thread.
+  create_process_promise_.GetFuture().Wait();
+  return std::move(process_info_or_error_);
+}
+
+void Debugger::Detach() { detach_requested_ = true; }
+
+void Debugger::Wait() {
+  if (thread_.joinable()) thread_.join();
+}
+
+void Debugger::DebuggerThread(std::filesystem::path executable,
+                              std::filesystem::path working_directory, std::string arguments) {
+  orbit_base::SetCurrentThreadName("OrbitDebugger");
+
+  // Only the thread that created the process being debugged can call WaitForDebugEvent.
+  process_info_or_error_ = CreateProcessToDebug(executable, working_directory, arguments);
+  create_process_promise_.SetResult(true);
```

Comment:
Using the Promise works, but it also means you won't benefit from thread sanitizer. If you replaced the promise by a mutex and wrapped the `process_info_or_error_` into a `std::optional` you would be able to use the `ABSL_GUARDED_BY_MUTEX` annotation.

The only culprit: Thread sanitizer doesn't work on Windows (yet?), so that's not really a strong argument.

Reply:
Good point, it doesn't apply after the change to `Promise<StartInfo>`.